File: 0542-01-matrix/0542-01-matrix.cpp
class Solution {
public:
    int findNearZero(vector<vector<int>> mat, int i, int j) {
        int level = 0;
        queue<pair<int, int>> q;
        q.push({i, j});
        
        mat[i][j] = -1;

        while (!q.empty()) {
            int n = q.size();
            for (int k = 0; k < n; k++) {
                pair<int, int> temp = q.front();
                q.pop();
                int x = temp.first;
                int y = temp.second;

                if (mat[x][y] == 0) {
                    return level;
                }

                if (x - 1 >= 0 && mat[x - 1][y] != -1) {
                    q.push({x - 1, y});
                    if (mat[x - 1][y] == 1) mat[x - 1][y] = -1;
                }
                if (y - 1 >= 0 && mat[x][y - 1] != -1) {
                    q.push({x, y - 1});
                    if (mat[x][y - 1] == 1) mat[x][y - 1] = -1;
                }
                if (x + 1 < mat.size() && mat[x + 1][y] != -1) {
                    q.push({x + 1, y});
                    if (mat[x + 1][y] == 1) mat[x + 1][y] = -1;
                }
                if (y + 1 < mat[0].size() && mat[x][y + 1] != -1) {
                    q.push({x, y + 1});
                    if (mat[x][y + 1] == 1) mat[x][y + 1] = -1;
                }
            }
            level++;
        }
        return level;
    }

    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int row = mat.size(), col = mat[0].size();
        vector<vector<int>> ans(row, vector<int>(col, 0));
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (mat[i][j] == 1) {
                    ans[i][j] = findNearZero(mat, i, j);
                }
            }
        }
        return ans;
    }
};

File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        q = deque()
        row, col = len(mat), len(mat[0])
        level = 1
        for i in range(row):
            for j in range(col):
                if mat[i][j] == 0:
                    q.append([i, j])
                else:
                    mat[i][j] = -1

        dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]

        while q:
            n = len(q)
            while n:
                temp = q.popleft()
                for dir in dirs:
                    i = temp[0] + dir[0]
                    j = temp[1] + dir[1]
                    if i < 0 or j < 0 or i >= row or j >= col or mat[i][j] != -1:
                        continue
                    mat[i][j] = level
                    q.append([i, j])
                n -= 1
            level += 1
        
        return mat